pulsar-broker/src/test/java/org/apache/pulsar/broker/service/MessageTTLTest.java

```diff
@@ -0,0 +1,107 @@
+import org.apache.pulsar.common.util.FutureUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class MessageTTLTest extends BrokerTestBase {
+
+    private static final Logger log = LoggerFactory.getLogger(MessageTTLTest.class);
+    @BeforeClass
+    @Override
+    protected void setup() throws Exception {
+        this.conf.setTtlDurationDefaultInSeconds(1);
+        this.conf.setBrokerDeleteInactiveTopicsEnabled(false);
+        super.baseSetup();
+    }
+
+    @AfterClass(alwaysRun = true)
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @Test
+    public void testMessageExpiryAfterTopicUnload() throws Exception {
+        int numMsgs = 50;
+        final String topicName = "persistent://prop/ns-abc/testttl";
+        final String subscriptionName = "ttl-sub-1";
+
+        pulsarClient.newConsumer()
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .subscribe()
+                .close();
+
+        Producer<byte[]> producer = pulsarClient.newProducer().topic(topicName)
+                .enableBatching(false) // this makes the test easier and predictable
+                .create();
+
+        List<CompletableFuture<MessageId>> sendFutureList = Lists.newArrayList();
+        for (int i = 0; i < numMsgs; i++) {
+            byte[] message = ("my-message-" + i).getBytes();
+            sendFutureList.add(producer.sendAsync(message));
+        }
+        FutureUtil.waitForAll(sendFutureList).get();
+        producer.close();
+        MessageId lastMessageId = sendFutureList.get(sendFutureList.size() - 1).get();
+
+        // unload a reload the topic
+        // this action created a new ledger
+        // having a managed ledger with more than one
+        // ledger should not impact message expiration
+        admin.topics().unload(topicName);
+        admin.topics().getStats(topicName);
+
+        AbstractTopic topic = (AbstractTopic) pulsar.getBrokerService().getTopicReference(topicName).get();
+        // wall clock time, we have to make the message to be considered "expired"
+        Thread.sleep(this.conf.getTtlDurationDefaultInSeconds() * 2000);
+        log.info("***** run message expiry now");
+        this.runMessageExpiryCheck();
+
+        Consumer<byte[]> consumer = pulsarClient.newConsumer()
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .subscribe();
+
+        MessageId lastMessageIdOnConsumer = consumer.getLastMessageId();
+        log.info("lastMessageID written {}, lastMessageIdForConsumer {}", lastMessageId, lastMessageIdOnConsumer);
```

Comment:
these two values are the same, even if the messages are expired

Reply:
You are right. I think you can get stats-internal to get the internal stats to get the markDeletePosition. If you see the markDeletePosition is moved, it means TTL is taking effect.

Reply:
very good @sijie 
I have updated the test